opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -376,46 +397,78 @@ def get_current_span(self) -> "Span":
         # pylint: disable=no-self-use
         return INVALID_SPAN
 
-    @contextmanager  # type: ignore
     def start_span(
         self,
         name: str,
         parent: ParentSpan = CURRENT_SPAN,
         kind: SpanKind = SpanKind.INTERNAL,
-    ) -> typing.Iterator["Span"]:
-        """Context manager for span creation.
+    ) -> "Span":
+        """Starts a span.
 
-        Create a new span. Start the span and set it as the current span in
-        this tracer's context.
+        Create a new span. Start the span without setting it as the current
+        span in this tracer's context.
 
         By default the current span will be used as parent, but an explicit
         parent can also be specified, either a `Span` or a `SpanContext`. If
         the specified value is `None`, the created span will be a root span.
 
-        On exiting the context manager stop the span and set its parent as the
+        The span can be used as context manager. On exiting, the span will be
+        ended.
+
+        Example::
+
+            # tracer.get_current_span() will be used as the implicit parent.
+            # If none is found, the created span will be a root instance.
+            with tracer.start_span("one") as child:
+                child.add_event("child's event")
+
+        Applications that need to set the newly created span as the current
+        instance should use :meth:`start_as_current_span` instead.
+
+        Args:
+            name: The name of the span to be created.
+            parent: The span's parent. Defaults to the current span.
+            kind: The span's kind (relationship to parent). Note that is
+                meaningful even if there is no parent.
+
+        Returns:
+            The newly-created span.
+        """
+        # pylint: disable=unused-argument,no-self-use
+        return INVALID_SPAN
+
+    def start_as_current_span(
+        self,
+        name: str,
+        parent: ParentSpan = CURRENT_SPAN,
+        kind: SpanKind = SpanKind.INTERNAL,
+    ) -> typing.Iterator["Span"]:
```

Comment:
Iterator is the wrong type here. Correct would be `ContextManger[Span]` but that does not exist in Python 3.5. So I guess this is a case where we have to use something like `-> SpanContextManager` and above

```python
MYPY = False
if MYPY:
    SpanContextManager = typing.ContextManager[Span]
else:
    SpanContextManager = typing.Any
```

https://mypy.readthedocs.io/en/latest/common_issues.html#import-cycles

Reply:
Honestly I'd leave this out for now, as it makes the code ugly ;)

(Else, I'd add it in another PR)

Reply:
I don't want to hold this PR up and I don't know why mypy accepts it and I agree that this code is very ugly. But: If I'm reading this correctly, the annotation is in fact wrong. If it were correct, you'd use this function as `for span in tracer().start_as_current_span()`. An iterator has no `set_attribute`, but it has `__next__`. A context manager has neither.

Reply:
I honestly feel like this is not the place to discuss this - this change it's even now, as we speak, in `master` - check out https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/trace/__init__.py#L385

-> It's still defined as `start_span` instead of `start_as_current_span` and in both cases, a `Span` is yielded.

I have no problem adjusting things that were changed in this PR, but I honestly see no reason to change things that were there before, unless they are fatal. Otherwise, we will in my next review something else that could be improved, and then we will have to discuss them and have yet another iteration ;)

Reply:
The point is that your removed the `@contextmanager` and consequently changed `yield` to `return` so it keeps working fine at runtime, but I think it messes up the type annotation. Unless the interaction between `@contextmanager` and the function return type is not taken into account by mypy, then it would already have been wrong before.

Reply:
--> #219